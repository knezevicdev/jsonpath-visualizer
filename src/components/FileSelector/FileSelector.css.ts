import styled from 'styled-components';

export const Input = styled.input`
  width: 0.1px;
  height: 0.1px;
  opacity: 0;
  overflow: hidden;
  position: absolute;
  z-index: -1;
`;

export const Label = styled.label`
  font-size: 1.25em;
  font-weight: 700;
  color: white;
  background-color: black;
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  padding: 10px 20px;

  ${Input}:focus + &,
  &:hover {
    background-color: red;
  }

  ${Input}:focus + & {
    outline: 1px dotted #000;
    outline: -webkit-focus-ring-color auto 5px;
  }

  & * {
    pointer-events: none;
  }

  @media (max-width: 768px) {
    span {
      display: none;
    }
  }
`;
