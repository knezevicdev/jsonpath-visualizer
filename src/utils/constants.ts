export const ARRAY_GROUP_LIMIT = 100;
