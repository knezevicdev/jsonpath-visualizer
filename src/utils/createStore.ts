export type TJson = null | Record<string, unknown> | any[];

export const createStore = () => {
  return {
    json: null as TJson,
    jsonPath: '' as string,
  };
};

export type TStore = ReturnType<typeof createStore>;
