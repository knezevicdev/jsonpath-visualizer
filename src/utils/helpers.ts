import { transform as _transform, isObject as _isObject, mapKeys as _mapKeys, assign as _assign } from 'lodash';
import { JSONPath } from 'jsonpath-plus';

import { TJson } from './store';

// eslint-disable-next-line
export const toType = (obj: any): string => {
  let type = getType(obj);
  if (type === 'number') {
    if (isNaN(obj)) {
      type = 'nan';
    } else if ((obj | 0) != obj) {
      //bitwise OR produces integers
      type = 'float';
    } else {
      type = 'integer';
    }
  }
  return type;
};

//source: http://stackoverflow.com/questions/7390426/better-way-to-get-type-of-a-javascript-variable/7390612#7390612
const getType = (obj: any): string => {
  return {}.toString
    .call(obj)
    .match(/\s([a-zA-Z]+)/)[1]
    .toLowerCase();
};

// eslint-disable-next-line
export const objectPaths = (obj: any): Record<string, unknown> => {
  return _transform(
    obj,
    (result: Record<string, unknown>, value: unknown, key: string) => {
      if (_isObject(value)) {
        const flatMap = _mapKeys(objectPaths(value), function (_mvalue, mkey) {
          return `${key}.${mkey}`;
        });

        _assign(result, flatMap);
      } else {
        result[String(key)] = false;
      }

      return result;
    },
    {},
  );
};

export const matchPaths = (json: TJson, path: string): string[] => {
  const paths = JSONPath({
    path,
    json,
    resultType: 'pointer',
  }).map((pointer: string) => pointer.replace(/\//g, '.').substring(1));

  if (paths.includes('')) {
    return [''];
  }

  return paths.reduce((matchedPaths: string[], path: string) => {
    for (const matchedPath in matchedPaths) {
      if (path.includes(matchedPath)) {
        return matchedPaths;
      }
    }

    return [...matchedPaths, path];
  }, []);
};
