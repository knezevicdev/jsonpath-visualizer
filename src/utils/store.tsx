import React, { ComponentType } from 'react';
import { useLocalStore } from 'mobx-react';
import { reaction, comparer, keys, remove } from 'mobx';

import { matchPaths } from 'utils/helpers';

export type TJson = null | Record<string, unknown> | any[];

const storeContext = React.createContext<Record<string, any> | null>(null);

export const StoreProvider: React.FC = ({ children }) => {
  const store = useLocalStore<Record<string, any>>(() => ({
    json: null as TJson,
    jsonPath: '' as string,
    jsonUploadTime: 0 as number,
    matchedPaths: [] as string[],
  }));

  reaction(
    () => store.jsonPath,
    (jsonPath) => {
      const matchedPaths = matchPaths(store.json, jsonPath);

      store.matchedPaths.forEach((oldMatchedPath: string) => {
        if (!matchedPaths.includes(oldMatchedPath)) {
          store[oldMatchedPath] = false;
        }
      });

      matchedPaths.forEach((matchedPath: string) => {
        if (!store[matchedPath]) {
          store[matchedPath] = true;
        }
      });

      store.matchedPaths = matchedPaths;
    },
  );

  reaction(
    () => store.json,
    () => {
      keys(store).forEach((key: string) => {
        if (!['json', 'jsonPath', 'matchedPaths', 'jsonUploadTime'].includes(key)) {
          remove(store, key);
        }
      });

      store.matchedPaths = [];
      store.jsonPath = '';
    },
    { equals: comparer.shallow },
  );

  return <storeContext.Provider value={store}>{children}</storeContext.Provider>;
};

export const withStore = (WrappedComponent: ComponentType): React.FC => {
  const StoreHOC: React.FC = (props) => {
    return (
      <StoreProvider>
        <WrappedComponent {...props} />
      </StoreProvider>
    );
  };

  return StoreHOC;
};

export const useStore = (): Record<string, any> => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
