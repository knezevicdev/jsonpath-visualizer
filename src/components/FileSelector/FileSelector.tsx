import React from 'react';
import { useToasts } from 'react-toast-notifications';
import { extendObservable } from 'mobx';
import { action } from 'mobx';

import { useStore, TJson } from 'utils/store';
import { toType, objectPaths } from 'utils/helpers';

import { Input, Label } from './FileSelector.css';

const FileSelector: React.FC = () => {
  const store = useStore();
  const { addToast } = useToasts();

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const reader = new FileReader();
    reader.onload = onReaderLoad;
    if (event.currentTarget.files?.length) reader.readAsText(event.currentTarget.files[0]);
  };

  const setJson = action((json: TJson) => {
    const paths = objectPaths(json);
    extendObservable(store, paths);

    store.json = json;
    store.jsonUploadTime = Date.now();
  });

  const onReaderLoad = (event: ProgressEvent<FileReader>): void => {
    const jsonString = event.target?.result;

    if (toType(jsonString) === 'string') {
      try {
        const result = JSON.parse(jsonString as string);

        setJson(result);
      } catch (e) {
        addToast('Please select valid JSON file', { appearance: 'error' });
      }
    }
  };

  return (
    <>
      <Input type="file" onChange={onChange} id="file-selector" />
      <Label htmlFor="file-selector">
        <i className="material-icons">file_upload</i>
        <span>&nbsp;Choose a file</span>
      </Label>
    </>
  );
};

export default FileSelector;
