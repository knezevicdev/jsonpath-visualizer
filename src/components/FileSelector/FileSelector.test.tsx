import React from 'react';
import { shallow } from 'enzyme';

import FileSelector from './FileSelector';
import { StoreProvider } from 'utils/store';

describe('FileSelector', () => {
  it('renders snapshots', () => {
    const wrapper = shallow(
      <StoreProvider>
        <FileSelector />
      </StoreProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
