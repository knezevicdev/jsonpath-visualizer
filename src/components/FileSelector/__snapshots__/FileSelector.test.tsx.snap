// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSelector renders snapshots 1`] = `
<ContextProvider
  value={
    Object {
      "json": null,
      "jsonPath": "",
      "jsonUploadTime": 0,
      "matchedPaths": Array [],
    }
  }
>
  <FileSelector />
</ContextProvider>
`;
